cmd/skopeo/signing.go

```diff
@@ -98,17 +104,34 @@ func (opts *standaloneVerifyOptions) run(args []string, stdout io.Writer) error
 		return fmt.Errorf("Error reading signature from %s: %v", signaturePath, err)
 	}
 
-	mech, err := signature.NewGPGSigningMechanism()
+	var mech signature.SigningMechanism
+	if opts.truststore != "" {
+		truststore, err := os.ReadFile(opts.truststore)
+		if err != nil {
+			return fmt.Errorf("Error reading trust store from %s: %v", opts.truststore, err)
+		}
+		mech, _, err = signature.NewEphemeralGPGSigningMechanism(truststore)
+	} else {
+		mech, err = signature.NewGPGSigningMechanism()
+	}
 	if err != nil {
 		return fmt.Errorf("Error initializing GPG: %v", err)
 	}
 	defer mech.Close()
+
+	if expectedFingerprint == "any" {
+		_, expectedFingerprint, err = mech.Verify(unverifiedSignature)
```

Comment:
Absolutely not. The fingerprint must come from a trusted source, not from the potentially-hostile signature. 

(And we should not accept keys of all humans that could possibly be imported into `$HOME/.gnupg`.)

Even if it might work out fine with the code as is, it’s a risk and the conditions that make it work now could be lost in future modifications.

The data is already available from a trusted source for ephemeral mechanisms.

Reply:
It's really intended to use with the new argument, to say "accept anything from this store".

If the key isn't in the store the verification fails, even with any.

Could make it only work if the argument is set?

Reply:
Conceptually, sure, if the public keys are explicitly provided, that already specifies the fingerprints, and `policy.json` uses the key data that way already. I just don’t want _this particular implementation approach_, reading the fingerprint from the untrusted signature, to be used; I’ve seen vulnerabilities from somewhat similar uses of untrusted data that are (originally) safe in context.

As for implementing this, `VerifyDockerManifestSignature` currently does not expose that feature to external callers. So we would need a small addition to c/image, which is, in isolation, not a big deal.

… but then again, why is the application not using `policy.json` already?

Reply:
Would you be more comfortable taking each of the fingerprints in the public key file and attempting verification in turn to see if one of them matches?

I'm not aware of a way to fake a fingerprint in a signature, but if that does exist, testing with a series of explicit keys wouldn't help, would it?

Reply:
I'll leave the policy.json question to the main thread :)

Reply:
I think a possible implementation (which is similar to what `policy.json` does, simplifying things)
- `signature.NewEphemeralGPGSigningMechanism` already returns the fingerprints that can be matched; those can be used instead of being ignored.
-  The core verification logic underlying `VerifyDockerManifestSignature` is already flexible WRT deciding which keys are acceptable; so it’s just a matter of exposing that in c/image, `VerifyImageManifestSignatureMultipleKeys` or something.

---

> I'm not aware of a way to fake a fingerprint in a signature

(BTW the signature doesn’t even contain the full fingerprint… just a trailing part that is used to look up a local key.)

> , but if that does exist, testing with a series of explicit keys wouldn't help, would it?

There’s nothing insecure about the code as it is _in this PR_. I’m just a touch paranoid :)

The thing I’m worried about is that it muddles the trust dependency relationship.

Before: fully trusted user → public keys and a specific fingerprint → cryptographic signature → signature payload → manifest. There’s a clear “source of trust → originally untrusted, newly trusted data” direction, and untrusted data is not used for any purpose until the prerequisites are verified.

After, in _some_ cases: fully trusted user → public keys → cryptographic signature → fingerprint → cryptographic signature → signature payload …

i.e. this line is fine on the code path where it is, but it would be problematic on the other code path, unexpectedly accepting any signature from $HOME/.gnupg ; or it could be copy&pasted into some other software…

Reply:
> The core verification logic underlying VerifyDockerManifestSignature is already flexible WRT deciding which keys are acceptable; so it’s just a matter of exposing that in c/image, VerifyImageManifestSignatureMultipleKeys or something.

I was hoping not to have to do the c/image dance if I could avoid it, which is why I went with the approach of verifying we have the key in the public key store, then running the proper verification.

I could just run the verification once per key without having to change c/image, but if you think changing c/image is the only acceptable approach, then I'll go add the new method and suck it up :)

Reply:
Actually, it would be hard to run it once per key cleanly because the error type is the same for all errors and I can't tell cleanly if its a fingerprint mismatch.

So, its either (a) the approach I have, but restrict it to only working if we specify a key file, or (b) update c/image with a multi-keyprint method.

I'd prefer to do (a), but if you think (b) is the only way, then I'll do that instead.

Reply:
Have changed the argument name and added a double-check of the fingerprint being found in the public key file. I might pass this around some colleagues so wanted the externals to look about right.

Reply:
Yes, I’d much prefer a new function in c/image , to make all the enforcement logic centralized … and yes, with a fairly stringent unit-test coverage requirements. Apart from `skopeo sync`, Skopeo is mostly intended to be a _very dumb_ wrapper around c/image.

(We don’t need a c/image release to add the Skopeo feature; Skopeo can just vendor a mainline commit.)

Reply:
OK, will do that next :)

The painful part last time was needing to make the skopeo tests pass in the c/Image PR. This time, I expect I don't need to do that bit.

Reply:
image pr: https://github.com/containers/image/pull/1904